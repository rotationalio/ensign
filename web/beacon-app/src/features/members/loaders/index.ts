export { default as userLoader } from './userLoader';
