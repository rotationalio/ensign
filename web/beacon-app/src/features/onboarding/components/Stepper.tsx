const Stepper = () => {
  return (
    <div>
      <h1>Stepper</h1>
    </div>
  );
};

export default Stepper;
