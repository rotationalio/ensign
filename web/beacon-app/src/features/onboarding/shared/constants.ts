export enum ONBOARDING_STEPS {
  ORGANIZATION = 1,
  WORKSPACE = 2,
  NAME = 3,
  PREFERENCE = 4,
  COMPLETE = 5,
}

export const WORKSPACE_DOMAIN_BASE = 'https://rotational.app/';
