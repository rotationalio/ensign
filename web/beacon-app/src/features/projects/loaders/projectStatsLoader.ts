import { json } from 'react-router-dom';

import axiosInstance from '@/application/api/ApiService';
import { queryClient } from '@/application/config/react-query';
import { tenantsRequest } from '@/features/tenants/api/tenantListAPI';

import { projectStatsApiRequest } from '../api/projectStatsApiService';

export async function projectPageLoader() {
  const tenants = await queryClient.fetchQuery({
    queryFn: tenantsRequest(axiosInstance),
  });

  const projectStats = await projectStatsApiRequest(axiosInstance)(tenants.tenants[0].id);

  return json({
    projectStats,
  });
}
