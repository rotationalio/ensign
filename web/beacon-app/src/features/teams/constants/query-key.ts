export const QUERY_KEY = {
  MEMBERS_LIST: 'memberList',
} as const;
