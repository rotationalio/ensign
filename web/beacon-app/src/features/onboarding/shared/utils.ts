import { t } from '@lingui/macro';

import { MemberResponse } from '@/features/members/types/memberServices';

import { ONBOARDING_STEPS } from './constants';
export const getDeveloperOptions = () => {
  return [
    { value: 'Application development', label: t`Application development` },
    { value: 'Data science', label: t`Data science` },
    { value: 'Data engineering', label: t`Data engineering` },
    { value: 'Developer experience', label: t`Developer experience` },
    { value: 'Cybersecurity (blue or purple team)', label: t`Cybersecurity (blue or purple team)` },
    { value: 'DevOps and observability', label: t`DevOps and observability` },

    { value: 'Something else', label: 'Something else' },
  ];
};

export const getProfessionOptions = () => [
  {
    id: 'work_segment',
    value: 'work',
    label: t`Work`,
  },
  {
    id: 'education_segment',
    value: 'education',
    label: t`Education`,
  },
  {
    id: 'personal_segment',
    value: 'personal',
    label: t`Personal`,
  },
];

export const getCurrentStepFromMember = (member: any) => {
  let current = 1;
  const hasOrganization = member?.organization?.length > 0;
  const hasWorkspace = member?.workspace?.length > 0;
  const hasName = member?.name?.length > 0;
  const hasProfessionSegment = member?.profession_segment?.length > 0;
  const hasDeveloperSegment = member?.developer_segment?.length > 0;
  const hasReachStep = (step: number) => {
    switch (step) {
      case 1:
        return !hasOrganization;
      case 2:
        return hasOrganization && !hasWorkspace;
      case 3:
        return hasOrganization && hasWorkspace && !hasName;
      case 4:
        return hasOrganization && hasWorkspace && hasName;
      default:
        return false;
    }
  };

  if (hasReachStep(ONBOARDING_STEPS.ORGANIZATION)) {
    current = 1;
  }

  if (hasReachStep(ONBOARDING_STEPS.WORKSPACE)) {
    current = 2;
  }

  if (hasReachStep(ONBOARDING_STEPS.NAME)) {
    current = 3;
  }

  if (hasReachStep(ONBOARDING_STEPS.PREFERENCE)) {
    current = 4;
  }

  if (
    !hasOrganization &&
    !hasWorkspace &&
    !hasName &&
    (!hasProfessionSegment || !hasDeveloperSegment)
  ) {
    current = 1;
  }

  return current;
};

export const getOnboardingStepsData = (member: Partial<MemberResponse>) => {
  const { name, organization, workspace, profession_segment, developer_segment } = member;

  return {
    name,
    organization,
    workspace,
    developer_segment,
    profession_segment,
  };
};

export const hasCompletedOnboarding = (member: MemberResponse) => {
  const { name, organization, workspace, profession_segment, developer_segment } = member;

  return name && organization && workspace && profession_segment.length > 0 && developer_segment;
};

export const isInvitedUser = (member: Pick<MemberResponse, 'invited'>) => {
  return member?.invited === true;
};

export const stepperContents = [
  {
    title: t`Step 1 of 4`,
    description: t`Your Team Name`,
  },
  {
    title: t`Step 2 of 4`,
    description: t`Your Workspace URL`,
  },
  {
    title: t`Step 3 of 4`,
    description: t`Your Name`,
  },
  {
    title: t`Step 4 of 4`,
    description: t`Your Goals`,
  },
];
