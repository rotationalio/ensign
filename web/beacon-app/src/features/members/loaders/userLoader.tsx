import { json } from 'react-router-dom';

import { queryClient } from '@/application/config/react-query';
import { useOrgStore } from '@/store';
import { clearCookies } from '@/utils/cookies';

import { memberDetailQuery } from '../hooks/useFetchMember';

const userLoader = () => async () => {
  const orgDataState = useOrgStore.getState() as any;
  const { user } = orgDataState || null;
  // console.log('userLoader', user);

  if (!user) {
    return null;
  } else {
    try {
      const query = memberDetailQuery(user);

      const member =
        queryClient.getQueryData(query?.queryKey) ?? (await queryClient.fetchQuery(query));
      return json({
        member,
      });
    } catch (error: any) {
      if (error?.response?.status === 401) {
        clearCookies();
        // routeLoader.clearCacheAll(); // TODO: clear cache all routes when logout
        window.location.href = '/';
      }
    }
  }
};

export default userLoader;
