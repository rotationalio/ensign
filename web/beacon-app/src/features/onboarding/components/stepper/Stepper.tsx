import { t } from '@lingui/macro';

import { useFetchProfile } from '@/features/members/hooks/useFetchProfile';

import { WORKSPACE_DOMAIN_BASE } from '../../shared/constants';
import { stepperContents } from '../../shared/utils';
import StepperStep from './StepperStep';

const Stepper = () => {
  const { profile: userProfile } = useFetchProfile();
  const isInvitedUser = userProfile?.invited;

  return (
    <>
      <ol className="stepper-items relative border-l border-gray-200 text-white">
        {stepperContents.map((step, idx) => (
          <StepperStep
            title={step.title}
            description={
              isInvitedUser && idx === 0
                ? t`Organization`
                : isInvitedUser && idx === 1
                ? t`Workspace URL`
                : step.description
            }
            value={
              isInvitedUser && idx === 0
                ? userProfile?.organization
                : isInvitedUser && idx === 1
                ? `${WORKSPACE_DOMAIN_BASE}${userProfile?.workspace}`
                : undefined
            }
            key={idx}
          />
        ))}
      </ol>
    </>
  );
};

export default Stepper;
