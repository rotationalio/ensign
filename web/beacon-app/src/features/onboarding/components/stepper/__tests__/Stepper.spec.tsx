import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render, screen } from '@testing-library/react';
import { vi } from 'vitest';

import * as userLoader from '@/features/members/loaders/userLoader';

import { WORKSPACE_DOMAIN_BASE } from '../../../shared/constants';
import Stepper from '../Stepper';
const renderComponent = () => {
  const queryClient = new QueryClient();
  const wrapper = ({ children }: any) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );
  return render(<Stepper />, { wrapper });
};

// mock t tag from lingui
vi.mock('@lingui/macro', () => ({
  t: (str) => str,
}));

vi.mock('@/features/members/loaders/userLoader', () => ({
  __esModule: true,
  default: () => ({
    member: {
      invited: false,
      organization: 'test',
      workspace: 'test',
    },
  }),
}));

describe('Stepper', () => {
  beforeEach(() => {
    vi.resetAllMocks();
  });

  it('should render the component with default value', () => {
    vi.spyOn(userLoader, 'default')
      .mockImplementation()
      .mockReturnValue({
        member: {
          invited: false,
          organization: 'test',
          workspace: 'test',
        },
      });

    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render the component for invited user', () => {
    const MockMember = {
      invited: true,
      organization: 'invited organization',
      workspace: 'invited-workspace',
    };

    vi.spyOn(userLoader, 'default')
      .mockImplementation()
      .mockReturnValue({
        member: {
          ...MockMember,
        },
      });

    renderComponent();

    expect(screen.getByText('Organization')).toBeInTheDocument();
    expect(screen.getByText('invited organization')).toBeInTheDocument();

    expect(screen.getByText('Workspace URL')).toBeInTheDocument();
    expect(screen.getByText(`${WORKSPACE_DOMAIN_BASE}${MockMember.workspace}`)).toBeInTheDocument();
  });
});
