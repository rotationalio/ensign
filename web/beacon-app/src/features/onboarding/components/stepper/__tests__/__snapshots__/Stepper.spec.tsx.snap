// Vitest Snapshot v1

exports[`Stepper > should render the component with default value 1`] = `
<div>
  <ol
    class="stepper-items relative border-l border-gray-200 text-white"
  >
    <li
      class="mb-10 ml-6"
    >
      <span
        class="stepper-item absolute -left-[4px] mt-1 flex h-2 w-2 items-center justify-center rounded-full bg-gray-100 ring-4 ring-white"
      />
      <h3
        class="font-medium leading-tight"
      >
        Step 1 of 4
      </h3>
      <p
        class="text-sm "
      >
        Your Team Name
      </p>
    </li>
    <li
      class="mb-10 ml-6"
    >
      <span
        class="stepper-item absolute -left-[4px] mt-1 flex h-2 w-2 items-center justify-center rounded-full bg-gray-100 ring-4 ring-white"
      />
      <h3
        class="font-medium leading-tight"
      >
        Step 2 of 4
      </h3>
      <p
        class="text-sm "
      >
        Your Workspace URL
      </p>
    </li>
    <li
      class="mb-10 ml-6"
    >
      <span
        class="stepper-item absolute -left-[4px] mt-1 flex h-2 w-2 items-center justify-center rounded-full bg-gray-100 ring-4 ring-white"
      />
      <h3
        class="font-medium leading-tight"
      >
        Step 3 of 4
      </h3>
      <p
        class="text-sm "
      >
        Your Name
      </p>
    </li>
    <li
      class="mb-10 ml-6"
    >
      <span
        class="stepper-item absolute -left-[4px] mt-1 flex h-2 w-2 items-center justify-center rounded-full bg-gray-100 ring-4 ring-white"
      />
      <h3
        class="font-medium leading-tight"
      >
        Step 4 of 4
      </h3>
      <p
        class="text-sm "
      >
        Your Goals
      </p>
    </li>
  </ol>
</div>
`;
